src/rebar_config.erl

```diff
@@ -92,8 +92,17 @@ consult_lock_file(File) ->
 %% at most once.
 %% The warning can also be cancelled by configuring the `warn_config_vsn'
 %% OTP env variable.
--spec warn_vsn_once() -> ok.
-warn_vsn_once() ->
+-spec warn_vsn_once(string()) -> ok.
```

Comment:
Initial ugly version, some de-duplication should happen here. @ferd @tsloughter I think for this case we should print the message and exit gracefully. There is no successful continuation for this case, but maybe trying to exit gracefully isn't worth it.

Reply:
The message also could be much better

Reply:
Question: Why do we have a semver style lock version? I would think an single integer would suffice, but I can also see where having the ability to maintain multiple maintenance versions of rebar3 would serve well from such a version schema.

Reply:
I just went with the most descriptive version possible when I started supporting them. I had no idea what would be needed in the future. For example, we could eventually decide that patch versions are not worth warnings because we'd be changing formatting or something of the kind, adding details that were optional but sometimes missing, and so on. Otherwise the Major/Minor version should let a theoretical Rebar4 know it found the lockfile of a Rebar3 and know it's not meant for it (or go read in some legacy mode).

In the case here we could have argued that the change should go for a new file format version (2.0.0) since we're moving from inner to outer hash without changing the key, and so the meaning is incompatible. However, if the `{pkg_hash, ...}` attribute currently used in this PR was replaced with `{pkg_ext_hash, ...}` we could probably keep the 1.2.0 version: a tool that analyzes all the hashfiles could do so without having to care about which it gets. If we wanted to play real nice we'd also keep the old `pkg_hash` value so that technically we would warn in older rebar3 versions, but wouldn't need to actually error out on the file. It would make people's diffs a bit larger but the transition neater. We could probably get rid of the `pkg_hash` key later down the line if ever needed after a few releases of `pkg_ext_hash`-supporting versions.

Reply:
To be clear, `pkg_hash` would contain inner checksums and `pkg_ext_hash` would contain outer?  That makes sense to me. Do we want to work this into rebar_pkg_resource so that people are not forced into unlocking and re-locking per upgrading to the latest version of rebar?

Reply:
Yeah; pkg_hash same as before, pkg_ext_hash the external checksums. We currently rewrite all lockfiles even if they're unchanged so as long as you write the new format down you should be fine -- Tristan has a PR to correct that, but the version upgrade is precisely the problem we're hitting our noses against right now in there.

Internally you can keep checking only the external hash, but this gives us a non-breakable way to keep working with older rebar3 copies.

Reply:
Ahhh I gotcha yeah. Will have this ready to go this evening.

Reply:
To note: This will need a little bit more ❤️ than anticipated. We have to update more modules than expected. Specifically, we are having to extend the app info tuple to contain both the new hash and the old hash (outer and inner). Most of the work is done now but have a few minor details to finish up tomorrow.